walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,135 @@
+        occupiedTimeSlots.remove(timeSlot);
+        addTimeRange(occupiedTimeSlots, addedTimeSlot);
+        return occupiedTimeSlots;
+      }
+      // New TimeRange overlaps with an already existing meeting.
+      if(timeSlot.overlaps(addedTimeSlot)) {
+        addTimeRange(occupiedTimeSlots, mergeTimeRanges(timeSlot, addedTimeSlot));
+        return occupiedTimeSlots;
+      }
+    }
+    // No scenario applied to TimeRange added normally.
+    occupiedTimeSlots.add(addedTimeSlot);
+    return occupiedTimeSlots;
+  }
+
+  /*
+   * Merges overlapping TimeRanges into a single TimeRange that includes both of them.
+   */
+  private TimeRange mergeTimeRanges(TimeRange a, TimeRange b) {
+    return TimeRange.fromStartEnd(a.start() < b.start() ? a.start() : b.start(),
+        a.end() > b.end() ? a.end() : b.end(), false);
+  }
+
+  /*
+   * Returns a collection with TimeRanges that are not available for meetings.
+   */
+  private ArrayList<TimeRange> getOccupiedTimeSlots(Collection<Event> events,
+      HashSet<String> requestAttendees) {
+    ArrayList<TimeRange> timeRangeEventsRequestAttendees = new ArrayList<TimeRange>();
+    for(Event meeting : events) {
+      for(String attendee : meeting.getAttendees()) {
+        // If the meeting contains a requested attendee it is added.
+        if(requestAttendees.contains(attendee) != false) {
+          timeRangeEventsRequestAttendees = addTimeRange(timeRangeEventsRequestAttendees,
+              meeting.getWhen());
+          break;
+        }
+      }
+    }
+    Collections.sort(timeRangeEventsRequestAttendees, TimeRange.ORDER_BY_START);
+    return timeRangeEventsRequestAttendees;
+  }
+
+  /*
+   * Calculates whether requested meetings's duration fits in between occupied TimeRanges.
+   */
+  private Collection<TimeRange> getFreeTimeSlots(ArrayList<TimeRange> occupiedTimeSlots, 
+      long duration) {
+    Collection<TimeRange> possibleTimeSlots = new ArrayList<>();
+    // Return all available if there are no occupied TimeRanges.
+    if(occupiedTimeSlots.isEmpty()) {
+      possibleTimeSlots.add(TimeRange.WHOLE_DAY);
+      return possibleTimeSlots;
+    }
+    // From midnight to the first meeting.
+    if(occupiedTimeSlots.get(0).start() - TimeRange.START_OF_DAY >= duration) {
+      possibleTimeSlots.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, occupiedTimeSlots.get(0).start(), false));
+    }
+    int previousEnd = occupiedTimeSlots.get(0).end();
+    int currentStart;
+    // Check for spaces between occupied TimeRanges.
+    for(int itr = 1; itr < occupiedTimeSlots.size(); itr++) {
+      currentStart = occupiedTimeSlots.get(itr).start();
+      int diff = currentStart - previousEnd;
+      if(currentStart - previousEnd >= duration) {
+        possibleTimeSlots.add(TimeRange.fromStartEnd(previousEnd, currentStart, false));
+      }
+      previousEnd = occupiedTimeSlots.get(itr).end();
+    }
+    // From last meeting to right before meeting.
```

Comment:
what does 'right before meeting' mean? Do you mean midnight?

Reply:
Ur right sorry for the confusion!